luarocks.lock

```diff
@@ -0,0 +1,14 @@
+return {
+   dependencies = {
+      dkjson = "2.6-1",
+      lua = "5.4-1",
```

Comment:
I understand most of the pinning, but one reason I've held off including a lock file in more projects is this Lua interpreter version pinning. Many/most projects I work on are explicitly designed and tested to run on multiple versions of Lua.

Is it possible to make `luarocks` cooperate with pinning everything except Lua itself? Does using the lock file still work if this line is dropped or is it all-or-nothing?

Reply:
> Many/most projects I work on are explicitly designed and tested to run on multiple versions of Lua.

This is something new to me, but I already saw it yesterday. Also when did check the lua release cycle on the site, I found almost all the old versions are kind of unsupported (the last release was long time ago), I am bit confused about lua's release lifecycle.

> Is it possible to make `luarocks` cooperate with pinning everything except Lua itself? Does using the lock file still work if this line is dropped or is it all-or-nothing?

I am bit surprised that `luarocks` lock `lua` version though, it can be outside as engine constraint (like npm, or other build tools). But for what I see in the other build tools, it is all or nothing.

I am gonna search a bit and see if anyone feedback on this `lua` line.

Reply:
Lua is a pretty fragmented ecosyntem with no single canonical direction. The main issue is that even though development continues in one direction (call this where Lua 5.4 is), it also got forked back in 5.1 to create LuaJIT. Since the JIT fork is substantially faster it has kept up a life of it's own with new features that diverged even from 5.1. Even that JIT project has been forked with another popular project (OpenResty) bringing another mix of features and tradeoffs. Aditionally Lua runtimes are embedded in lots of other software (e.g. Pandoc), ofter with version specific bindings. Complicating matters is the widespread use of distros with LTS releases that have old versions out there. Hence many projects in the Lua realm have to either pick a lane or deliberately use a lowest-common-denominator approach to run on as many as possible.

Busted itself *must* take the latter approach, especially since it gets used to test projects that are limited to various different interpreters.

Reply:
@hishamhm Any comment on why the VM is included in the `make --pin` output this way?

Reply:
Gonna wait for the comment before doing anything.

Reply:
@alerque Because this is intended as a way to get more stable builds via a fixed set of dependencies — some rocks will fetch different dependencies based on Lua version, so a given dependency graph is actually dependent on the Lua version (for example, IIRC, some of lhf's modules have different versions for different Lua versions).